// config/models.js
var mongoose = require('mongoose');

var userSchema = mongoose.Schema({
    token: String,
    phone_number: String,
    code: String,
    vertify_flag: Boolean,
    hashed_password: String,
    sex: String,
    email: String,
    first_name: String,
    last_name: String,
    driver_flag: Boolean,
    driver_license: String,
    plate_number: String,
    colour: String,
    car_model: String,
    salt: String,
});

var tripSchema = mongoose.Schema({
    user_id: String,
    date_id: String,
    start_lat: String,
    start_lng: String,
    start_location: String,
    end_lat: String,
    end_lng: String,
    end_location: String,
    est_time: String,
    est_distance: String,
    reg_as: String,
    other_user: {type: Array, "default" : []}
});

var chatContactSchema = mongoose.Schema({
    my_user_id: String,
    other_user_id: String
});

var chatHistorySchema = mongoose.Schema({
    sender_user_id: String,
    receiver_user_id: String,
    message: String
});

mongoose.connect('mongodb://vgo:thomas@ec2-52-40-59-253.us-west-2.compute.amazonaws.com:27017/vgoDB');
var User = mongoose.model('user', userSchema);
var Trip = mongoose.model('trip', tripSchema);
var ChatContact = mongoose.model('chat_contact', chatContactSchema);
var ChatHistory = mongoose.model('chat_history', chatHistorySchema);

module.exports = {
    User: User,
    Trip: Trip,
    ChatHistory: ChatHistory,
    ChatContact: ChatContact
};
