// config/models.js
var mongoose = require('mongoose');

var userSchema = mongoose.Schema({
    token: String,
    phone_number: String,
    code: String,
    vertify_flag: Boolean,
    hashed_password: String,
    sex: String,
    email: String,
    first_name: String,
    last_name: String,
    driver_flag: Boolean,
    driver_license: String,
    plate_number: String,
    colour: String,
    car_model: String,
    salt: String,
});

var chatContactSchema = mongoose.Schema({
    my_user_id: String,
    other_user_id: String,
    last_msg: String,
    last_msg_date_time: String,
    number_of_unread_msg: Number
});

var chatHistorySchema = mongoose.Schema({
    sender_user_id: String,
    receiver_user_id: String,
    message: String
});

mongoose.connect('mongodb://vgo:thomas@ec2-52-40-59-253.us-west-2.compute.amazonaws.com:27017/vgoDB');
var User = mongoose.model('user', userSchema);
var ChatContact = mongoose.model('chat_contact', chatContactSchema);
var ChatHistory = mongoose.model('chat_history', chatHistorySchema);

module.exports = {
    User: User,
    ChatHistory: ChatHistory,
    ChatContact: ChatContact
};
