// config/sendCode.js
/**
 * Created by Sven on 7/5/2016.
 */
/* PATH: host_url:8080/sendCode
 *
 * INPUT:
 * 'phone_number'
 *
 * OUTPUT:
 *  JSON Object that contains
 *  'code' : respond code
 *  'msg' : respond message
 *  'object_id' : upon successful send code and register phone in database(only for debug purposes)
 *
 *
 *  1 -> Successfully register phone in database
 * -1 -> Fail to send validation code to the user
 * -6 -> Incorrect phone_number format
 * -7 -> Required field not set
 */
var config = require('./config');
var twilio = require('twilio');
var client = new twilio.RestClient(config.accountSid, config.authToken);

var objectid = '';

var mongoose = require('mongoose');
var models = require('config/models');

exports.sendCode = function(phone_number,area_code,callback) {

    if (!phone_number || !area_code) {
        callback({
            'code': "-7",
            'msg': "Required field not set"
        });
        return;
    }

    if (phone_number.trim().length == 0 || phone_number.trim().length != 10 || !/^\d+$/.test(phone_number)) {
        callback({
            'code': "-6",
            'msg': "Incorrect phone_number format"
        });
        return;
    }

    var code = Math.floor(1000 + Math.random() * 9000);

    model.User.find({phone_number: phone_number},function(err,users){
        if(users.length == 0){

            console.log('This is a new user!');

            var newUser = new models.User ({
                phone_number: phone_number,
                code: code,
                validate: false
            });

            newUser.save(function(error){
                console.log('registered in database');
                model.User.find({phone_number: phone_number},function(err,users){
                    objectid = users[0]._id;
                })
                console.log('the user object_id is ' + objectid);
            })
        }else{
            console.log('This is a returning user!')
            user.findOne({phone_number: phone_number}, function (err,doc) {
                doc.code = code;
                doc.save();
            })
            objectid = users[0]._id;
            console.log('the user object_id is ' + objectid);
        }
    })

    client.sms.create({
        to: area_code + phone_number,
        from: config.twilioNumber,
        body: '[V-GO Group] Your verification code for validation is ' + code +' .'
    }, function (error, message) {
        // The HTTP request to Twilio will run asynchronously. This callback
        // function will be called when a response is received from Twilio
        // The "error" variable will contain error information, if any.
        // If the request was successful, this value will be "falsy"
        if (!error) {
            // The second argument to the callback will contain the information
            // sent back by Twilio for the request. In this case, it is the
            // information about the text messsage you just sent:
            console.log('Success! The SID for this SMS message is:');
            console.log(message.sid);

            console.log('Message sent to:');
            console.log(phone_number);

            console.log('Message sent on:');
            console.log(message.dateCreated);

            callback({
                'code': "1",
                'msg': "Successfully register phone in database, the objectid is "+ objectid
            });
            return;
        } else {
            console.log('Oops! There was an error.');
            console.log(error);
            callback({
                'code': "-1",
                'msg': "fail to send validation code to the user~"
            });
        }
    })
}
