// config/trip.js
/**
 * Created by Sven on 7/18/2016.
 */
var mongoose = require('mongoose');
var models = require('config/models');

exports.create = function(date_id, start_lat, start_lng, start_location, end_lat, end_lng, end_location, est_time, est_distance, reg_as, sess, callback) {
    var user_id_in_sess = sess.user_id
    if(!user_id_in_sess){
        callback({
            'code': "-",
            'msg': "No session, please login"
        });
        return;
    }

    if (!date_id || !start_lat || !start_lng || !start_location || !end_lat || !end_lng || !end_location || !est_time || !est_distance) {
        callback({
            'code': "-7",
            'msg': "Required field not set"
        });
        return;
    }
    
    var newTrip = new models.Trip({
        user_id: user_id_in_sess,
        date_id: date_id,
        start_lat: start_lat,
        start_lng: start_lng,
        start_location: start_location,
        end_lat: end_lat,
        end_lng: end_lng,
        end_location: end_location,
        est_time: est_time,
        est_distance: est_distance,
        reg_as: reg_as,
    });

    newTrip.save(function (error) {
        console.log('trip saved in database');
        callback({
            'code': "1",
            'msg': "Trip created successfully",
        });
    })
}

exports.getSelf = function(option, value, sess, callback){
    var user_id_in_sess = sess.user_id
    switch(option) {
        case 0:
            break;
        case 1:
            break;
        case 2:
            break;
        default:
            callback({
                'code': "-",
                'msg': "not a valid option"
            });
    }
}
